lib/resourceutil/util.go

```diff
@@ -0,0 +1,97 @@
+package resourceutil
+
+import (
+	"sort"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/syntasso/kratix/lib/hash"
+	"github.com/syntasso/kratix/lib/pipeline"
+	batchv1 "k8s.io/api/batch/v1"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+)
+
+const PipelineCompletedCondition = clusterv1.ConditionType("PipelineCompleted")
+
+func HasPipelineCompletedCondition(obj *unstructured.Unstructured) bool {
+	return HasCondition(obj, PipelineCompletedCondition)
+}
+
+func LastPipelineCompleted(obj *unstructured.Unstructured) bool {
+	condition := GetCondition(obj, PipelineCompletedCondition)
+	return condition != nil && condition.Status == v1.ConditionTrue
+}
+
+func MarkPipelineAsRunning(logger logr.Logger, obj *unstructured.Unstructured) {
+	SetCondition(obj, &clusterv1.Condition{
+		Type:               PipelineCompletedCondition,
+		Status:             v1.ConditionFalse,
+		Message:            "Pipeline has not completed",
+		Reason:             "PipelineNotCompleted",
+		LastTransitionTime: metav1.NewTime(time.Now()),
+	})
+	logger.Info("set conditions", "condition", PipelineCompletedCondition, "value", v1.ConditionFalse)
+}
+
+func MarkPipelineAsCompleted(logger logr.Logger, obj *unstructured.Unstructured) {
+	SetCondition(obj, &clusterv1.Condition{
+		Type:               PipelineCompletedCondition,
+		Status:             v1.ConditionTrue,
+		Message:            "Pipeline completed",
+		Reason:             "PipelineExecutedSuccessfully",
+		LastTransitionTime: metav1.NewTime(time.Now()),
+	})
+	logger.Info("set conditions", "condition", PipelineCompletedCondition, "value", v1.ConditionTrue)
+}
+
+func IsAnUpdate(logger logr.Logger, rr *unstructured.Unstructured, jobs []batchv1.Job) bool {
+	if len(jobs) == 0 {
+		return false
+	}
+
+	// sort the pipepeineJobs by creation date
+	sort.Slice(jobs, func(i, j int) bool {
+		t1 := jobs[i].GetCreationTimestamp().Time
+		t2 := jobs[j].GetCreationTimestamp().Time
+		return t1.Before(t2)
+	})
+
+	lastPipelineJob := jobs[len(jobs)-1]
+
+	lastRequestHash := lastPipelineJob.GetLabels()[pipeline.KratixResourceHashLabel]
+	currentRequestHash, err := hash.ComputeHash(rr)
+	if err != nil {
+		logger.Info("Cannot determine if the request is an update. Requeueing", "reason", err.Error())
+		return false
+	}
+
+	return lastRequestHash != currentRequestHash
+}
+
+func IsThereAPipelineRunning(logger logr.Logger, jobs []batchv1.Job) bool {
+	if len(jobs) == 0 {
+		return false
+	}
+
+	var incompleteJobs bool
+	for _, job := range jobs {
+		for _, condition := range job.Status.Conditions {
+			switch condition.Type {
+			// TODO: deal with other conditions; what should kratix do if
+			// the job is in a failed or suspended state?
```

Comment:
yeah this feels like we will get into a pretty frozen state with the code as is today, maybe if a Job is stalled/suspended we ignore it

Reply:
yeah right now is not gonna trigger an update until the pipeline completes successfully. i don't think we need to fix this now to merge this in though (we do need to think about it and fix it prior to "releasing")

Reply:
💯